### PR TITLE
error: handle bucket version not match

### PR DESCRIPTION
**Pull request by bufferflies**
https://github.com/tikv/tikv/pull/15224

### CODE REVIEW

File: internal/locate/region_request.go [hunk 1]
```diff
@@ -2038,6 +2038,17 @@ func (s *RegionRequestSender) onRegionError(
 		return retry, err
 	}
 
+	if bucketVersionNotMatch := regionErr.GetBucketVersionNotMatch(); bucketVersionNotMatch != nil {
+		logutil.BgLogger().Debug(
```
Comment (crazycs520) [inline internal/locate/region_request.go:2042]:
```suggestion
		logutil.Logger(bo.GetCtx()).Debug(
```

File: internal/locate/region_cache.go [hunk 2]
```diff
@@ -1947,6 +1948,27 @@ func (c *RegionCache) getStoresByLabels(labels []*metapb.StoreLabel) []*Store {
 	return s
 }
 
+// OnBucketVersionNotMatch removes the old buckets meta if the version is stale.
+func (c *RegionCache) OnBucketVersionNotMatch(ctx *RPCContext, version uint64, keys [][]byte) {
+	r := c.GetCachedRegionWithRLock(ctx.Region)
+	if r == nil {
+		println("reigon not found")
```
Comment (crazycs520) [inline internal/locate/region_cache.go:1955]: remove this debug code?

File: internal/locate/region_cache.go [hunk 3]
```diff
@@ -2095,6 +2117,26 @@ func (c *RegionCache) InvalidateTiFlashComputeStores() {
 	c.tiflashComputeStoreMu.needReload = true
 }
 
+// UpdateBucketOnError updates the buckets of the region in the cache if meets the buckets version not match.
+func (c *RegionCache) UpdateBucketOnError(regionID RegionVerID, version uint64, keys [][]byte) {
```
Comment (crazycs520) [inline internal/locate/region_cache.go:2121]: This function is never used? even not occur in the test.

**Review by crazycs520 — APPROVED** [review]

**Review by iosmanthus — APPROVED** [review]

**Review by nolouch — APPROVED** [review]

**Review by SpadeA-Tang — APPROVED** [review]

**Pull request merged**